src/main/java/net/snowflake/client/core/SFArrowResultSet.java

```diff
@@ -574,22 +575,66 @@ public Object getObject(int columnIndex) throws SFException {
     converter.setTreatNTZAsUTC(treatNTZAsUTC);
     converter.setUseSessionTimezone(useSessionTimezone);
     converter.setSessionTimeZone(sessionTimeZone);
-    Object obj = converter.toObject(index);
     boolean isStructuredType = resultSetMetaData.isStructuredTypeColumn(columnIndex);
-    if (isVarcharConvertedStruct(type, isStructuredType, converter)) {
-      if (obj != null) {
+    Object obj = converter.toObject(index);
+    if (obj == null) {
+      return null;
+    }
+    if (type == Types.STRUCT && isStructuredType) {
+      if (converter instanceof VarCharConverter) {
         return new StructObjectWrapper((String) obj, createJsonSqlInput(columnIndex, obj));
+      } else if (converter instanceof StructConverter) {
+        return new StructObjectWrapper(
+            converter.toString(index),
+            createArrowSqlInput(columnIndex, (Map<String, Object>) obj),
+            obj);
+      }
+    }
+    return new StructObjectWrapper(converter.toString(index), null, obj);
+  }
+
+  @Override
+  public <T> Object getObject(int columnIndex, Class<T> type) throws SFException {
+    if (type == String.class) {
+      return getObject(columnIndex);
+    }
+
+    int columnType = resultSetMetaData.getColumnType(columnIndex);
+    if (columnType == SnowflakeUtil.EXTRA_TYPES_VECTOR) {
+      return getString(columnIndex);
+    }
+    ArrowVectorConverter converter = currentChunkIterator.getCurrentConverter(columnIndex - 1);
+    int index = currentChunkIterator.getCurrentRowInRecordBatch();
+    wasNull = converter.isNull(index);
+    converter.setTreatNTZAsUTC(treatNTZAsUTC);
+    converter.setUseSessionTimezone(useSessionTimezone);
+    converter.setSessionTimeZone(sessionTimeZone);
+    Object obj = converter.toObject(index);
+    boolean isStructuredType = resultSetMetaData.isStructuredTypeColumn(columnIndex);
+    if (obj == null) {
+      return null;
+    }
```

Comment:
We have a lot of duplicated code lines between getObject(index) and getObject(index, type)

Reply:
moved some part of it to external method